psh/pm/pm.c

```diff
@@ -119,38 +175,63 @@ int psh_pm(int argc, char *argv[])
 	for (;;) {
 		sleep(interval);
 
-		ctcnt = psh_pm_getThreads(&ctinfo, &ctsize);
-		if (ctcnt < 0)
+		int ctcnt = psh_pm_getThreads(&ctinfo, &ctsize);
+		if (ctcnt < 0) {
+			if (rebootNoMem && ctcnt == -ENOMEM) {
+				psh_pm_reboot("ENOMEM while getting thread info");
+			}
 			continue;
+		}
 
-		i = 0;
-		j = 0;
+		uint32_t currKernel = 0;
+
+		int i = 0;
+		int j = 0;
 
 		while (i < itcnt) {
-			ipid = itinfo[i].pid;
-			cpid = ctinfo[j].pid;
+			pid_t ipid = itinfo[i].pid;
+			pid_t cpid = ctinfo[j].pid;
+
+			if (ctinfo[i].pid == 0) { /* kernel idle process - shows kernel RAM usage */
+				currKernel = ctinfo[i].vmem;
+			}
 
 			if ((!ignore_ppid || ipid != ppid) && (ipid < cpid)) {
 				fprintf(stderr, "pm: process %d died\n", ipid);
 				if (restart) {
-					err = reboot(PHOENIX_REBOOT_MAGIC);
-					if (err < 0)
-						fprintf(stderr, "pm: failed to restart the machine\n");
+					psh_pm_reboot("monitored process died");
 				}
 			}
 
 			if (ipid <= cpid) {
+				int k;
 				for (k = i + 1; k < itcnt && itinfo[k].pid == itinfo[i].pid; k++)
 					;
 				i = k;
 			}
 
 			if (ipid >= cpid) {
+				int k;
 				for (k = j + 1; k < ctcnt && ctinfo[k].pid == ctinfo[j].pid; k++)
 					;
 				j = k;
 			}
 		}
+
+		if (rebootNoMem) {
+			uint32_t currTotal = psh_pm_getTotal();
+			if (((5 * currTotal / 4) > maxTotal) || ((5 * currKernel / 4) > maxKernel)) {
```

Comment:
Why use 125% criterium? It's kinda misleading, as we input specific max amount

Reply:
Mega nitpick - will fail near UINT32_MAX curr*

Reply:
The idea was to print something into `dmesg` when we're nearing the limit (without using floats, maybe unnecessarily). The last message before reboot would obviously not make it to any remote client ;).

I could make this a condition only for printing not to impair the rebooting part in case of overflow (still, we're using int/unsigned int for counting vmem/memory in `meminfo_t`).

Please note that especially `threadinfo_t.vmem` is of type int, we should probably change it at least to unsigned.

Reply:
Ok, I misunderstood the code

Reply:
I also find it misleading. Maybe we could monitor and print max memory usage instead.

Reply:
I wanted to stick with something repetitive to have some information in `/var/log/kernel` if current  usage is dangerously high. I assume this is an error condition which should never happen in normal device work.

After some thoughts - maybe I can just use floats as all targets using psh do have FPU?

Reply:
To handle overflow you scale down max limits or introduce explicit option to set warning levels. I still think we should monitor kernel and total memory separately.

Reply:
I've introduced explicit `warn*` consts for better readability (and to add custom thresholds in the future if the need would arise)